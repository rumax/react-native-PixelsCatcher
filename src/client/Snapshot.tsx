/**
 * Copyright (c) Maksym Rusynyk 2018 - present
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// eslint-disable-next-line no-use-before-define
import React, { Component } from 'react';
import { InteractionManager, ScrollView } from 'react-native';

import log from './utils/log';

type SnapshotPropsType = { onReady: Function };

const TAG = 'PIXELS_CATCHER::APP::SNAPSHOT';
const ERROR_NO_IMPLEMENTED =
  'Not implemented. Should be implemented by actual snapshot';

export default class Snapshot extends Component<SnapshotPropsType, void> {
  // Should be implemented by actual snapshot
  static snapshotName: string = '';

  componentDidMount() {
    log.v(TAG, 'Awaiting interaction');
    const startTime = new Date().getTime();
    InteractionManager.runAfterInteractions(() => {
      const time = new Date().getTime() - startTime;
      log.v(TAG, `Interaction completed in ${time} milliseconds`);
      global.requestAnimationFrame(() => {
        this.props.onReady();
      });
    });
  }

  renderContent() {
    log.e(TAG, ERROR_NO_IMPLEMENTED);
    throw new Error(ERROR_NO_IMPLEMENTED);
  }

  render() {
    const content: any = this.renderContent();
    return (
      <ScrollView collapsable={false} contentContainerStyle={{ flexGrow: 1 }}>
        {content}
      </ScrollView>
    );
  }
}
