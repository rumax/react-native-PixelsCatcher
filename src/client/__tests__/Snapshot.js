/* @flow */
// eslint-disable-next-line no-use-before-define
import React from 'react';
import { View } from 'react-native';
import renderer from 'react-test-renderer';

import Snapshot from '../Snapshot';

jest.mock('../utils/log', () => ({ e: () => {}, v: () => {} }));
global.console.error = () => {};

describe('Snapshot component', () => {
  const onReadyMock = jest.fn();

  it('throws exception if renderContent is not implemented', () => {
    let exception;

    try {
      renderer.create(<Snapshot onReady={onReadyMock} />);
    } catch (err) {
      exception = err;
    }

    expect(exception).toMatchSnapshot();
  });

  it('renders snapshot component', () => {
    class SnapshotClass extends Snapshot {
      static snapshotName = 'AppSnapshot';

      renderContent() {
        return (
          <View />
        );
      }
    }

    const tree = renderer.create(<SnapshotClass onReady={onReadyMock} />);

    expect(tree).toMatchSnapshot();
  });
});
