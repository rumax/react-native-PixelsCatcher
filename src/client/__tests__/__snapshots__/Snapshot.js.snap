// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot component renders snapshot component 1`] = `
<RCTScrollView
  collapsable={false}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
>
  <View>
    <View />
  </View>
</RCTScrollView>
`;

exports[`Snapshot component renders snpashot component 1`] = `
<RCTScrollView
  collapsable={false}
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
>
  <View>
    <View />
  </View>
</RCTScrollView>
`;

exports[`Snapshot component throws exception if renderContent is not implemented 1`] = `[Error: Not implemented. Should be implemented by actual snapshot]`;
